sharktank/sharktank/utils/iree.py

```diff
@@ -72,39 +84,213 @@ def oneshot_iree_run(
+        save_vmfb_to: Optional path to save the compiled VMFB
+
+    Returns:
+        A TorchLikeIreeModule that can be called like the original torch module.
+        Single outputs are unwrapped, multiple outputs are returned as tuples.
+
+    Example:
+        >>> model = MyTorchModel()
+        >>> example_input = torch.randn(1, 3, 224, 224)
+        >>> iree_model = adapt_torch_module_to_iree(
+        ...     model,
+        ...     example_args=(example_input,),
+        ...     device="local-task"
+        ... )
+        >>> output = iree_model.forward(example_input)  # Single tensor, not list/tuple
+    """
+    vmfb_bytes = compile_torch_module_to_iree(
+        module=module,
+        example_args=example_args,
+        example_kwargs=example_kwargs,
+        function_name=function_name,
+        compile_args=compile_args,
+        save_mlir_to=save_mlir_to,
+        save_vmfb_to=save_vmfb_to,
+    )
+
     iree_devices = get_iree_devices(device=device, device_count=device_count)
 
-    def run(iree_devices: list[iree.runtime.HalDevice]):
+    def load_fn(devices: list[iree.runtime.HalDevice]) -> TorchLikeIreeModule:
         vm_module, vm_context, vm_instance = load_iree_module(
-            module_buff=memory_view, devices=iree_devices
-        )
-        torch_like_iree_module = TorchLikeIreeModule(
-            vm_module, vm_context, iree_devices
+            module_buff=vmfb_bytes,
+            devices=devices,
+            parameters_path=parameters_path,
+            tensor_parallel_size=len(devices),
         )
-        results = getattr(torch_like_iree_module, function)(*args, **kwargs)
-        # Clone to avoid leaking IREE-backed torch tensors.
-        results = tuple(t.clone() for t in results)
-        return results
+        return TorchLikeIreeModule(vm_module, vm_context, devices)
+
+    return with_iree_device_context(load_fn, iree_devices)
+
+
+@runtime_checkable
+class InferenceModule(Protocol):
+    """Protocol for inference modules (both torch and IREE).
+
+    This defines a common interface that both torch.nn.Module and
+    TorchLikeIreeModule can satisfy, allowing them to be used
+    interchangeably in inference code.
+
+    Example:
+        >>> def run_inference(model: InferenceModule, inputs):
+        ...     return model(inputs)
+        >>>
+        >>> # Works with torch modules
+        >>> torch_model = MyTorchModel()
+        >>> run_inference(torch_model, x)
+        >>>
+        >>> # Also works with IREE modules
+        >>> iree_model = adapt_torch_module_to_iree(torch_model, ...)
+        >>> run_inference(iree_model, x)
+    """
 
-    return with_iree_device_context(run, iree_devices)
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
+        """Execute the module's forward pass."""
+        ...
+
+    def forward(self, *args: Any, **kwargs: Any) -> Any:
+        """Execute the module's forward pass explicitly."""
+        ...
```

Comment:
We could have other functions, not just `forward`. Also there my be no `forward` at all.
The `@runtime_checkable` decorator would allow us to use `isinstance`, but if a module lacks a `forward` method, `isinstance` would return `False`.